File: 102.binary-tree-level-order-traversal/solution.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun levelOrder(root: TreeNode?): List<List<Int>> {
        val res: MutableList<List<Int>> = mutableListOf()
        val queue = ArrayDeque<TreeNode>()
        queue.offerUnlessNull(root)
        while (queue.isNotEmpty()) {
            val level = mutableListOf<Int>()
            repeat(queue.size) {
                val node = queue.poll()
                level.add(node.`val`)
                queue.offerUnlessNull(node.left)
                queue.offerUnlessNull(node.right)
            }
            res.add(level)
        }
        return res
    }
}

inline fun <T> ArrayDeque<T>.offerUnlessNull(elem: T?) {
    if (elem != null) offer(elem)
}

File: 102.binary-tree-level-order-traversal/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func levelOrder(root *TreeNode) [][]int {
    if root == nil {
        return [][]int{}
    }
    res := make([][]int, 0)
    queue := []*TreeNode{ root }
    for len(queue) > 0 {
        n := len(queue)
        level := make([]int, n)
        for i := 0; i < n; i++ {
            node := queue[i]
            level[i] = node.Val
            if node.Left != nil {
                queue = append(queue, node.Left)
            }
            if node.Right != nil {
                queue = append(queue, node.Right)
            }
        }
        res = append(res, level)
        queue = queue[n:]
    }
    return res
}